File: src/apply_style.ts
/**
 * Returns style attributes for a given class
 *
 *     data = { styles: { ':root': { color: 'gold', dir: 'none' } } }
 *     applyStyle(data, [':root'])
 *     => ['color="gold"', 'dir="none"']
 *
 * @param stylesheet the base stylesheet
 * @param data the data object with styles
 * @param classes the classes to apply
 * @param options optional before and after styles
 */
export default function applyStyle(
  stylesheet: Record<string, object>,
  data: { styles?: Record<string, object> },
  classes: string[],
  options?: { before?: object; after?: object }
) {
  if (!options) options = {};

  function applyStyles(acc: object, stylesheet: Record<string, object>) {
    return Object.keys(stylesheet).reduce((styles, key) => {
      if (classes.indexOf(key) === -1) return styles;
      return Object.assign({}, styles, stylesheet[key]);
    }, acc);
  }
  let result: object = {};
  result = Object.assign({}, result, options.before || {});
  result = applyStyles(result, stylesheet);
  result = applyStyles(result, data.styles || {});
  result = Object.assign({}, result, options.after || {});
  return renderStyle(result);
}

/**
 * Renders key/value into an attribute string
 *
 *     renderStyle({ color: 'gold', dir: 'none' ])
 *     => ['color="gold"', 'dir="none"']
 */
function renderStyle(properties: object) {
  return Object.entries(properties)
    .map(([key, val]) => {
      if (val == null) return null;
      return `${key}=${stringify(val)}`;
    })
    .filter((s) => s != null);
}

function stringify(val: string | object) {
  if (typeof val === 'string' && /^<.*>$/.test(val)) {
    return val;
  } else {
    return JSON.stringify(val);
  }
}

File: internal/render/apply_style.go
package render

import (
	"encoding/json"
	"regexp"
)

type applyOptions struct {
	Before map[string]any
	After  map[string]any
}

func ApplyStyle(stylesheet map[string]map[string]any, dataStyles map[string]map[string]any, classes []string, options *applyOptions) []string {
	if options == nil {
		options = &applyOptions{}
	}

	result := map[string]any{}
	for k, v := range options.Before {
		result[k] = v
	}

	result = applyStyles(result, stylesheet, classes)
	result = applyStyles(result, dataStyles, classes)

	for k, v := range options.After {
		result[k] = v
	}

	return renderStyle(result)
}

func applyStyles(acc map[string]any, stylesheet map[string]map[string]any, classes []string) map[string]any {
	if stylesheet == nil {
		return acc
	}
	classSet := map[string]bool{}
	for _, class := range classes {
		classSet[class] = true
	}

	for key, style := range stylesheet {
		if !classSet[key] {
			continue
		}
		for k, v := range style {
			acc[k] = v
		}
	}

	return acc
}

func renderStyle(properties map[string]any) []string {
	out := make([]string, 0, len(properties))
	for key, val := range properties {
		if val == nil {
			continue
		}
		out = append(out, key+"="+stringify(val))
	}
	return out
}

var htmlLike = regexp.MustCompile(`^<.*>$`)

func stringify(val any) string {
	if s, ok := val.(string); ok && htmlLike.MatchString(s) {
		return s
	}
	raw, _ := json.Marshal(val)
	return string(raw)
}
